trie/hexary.py

```diff
@@ -535,6 +542,14 @@ def squash_changes(self):
             yield memory_trie
         self.root_node = memory_trie.root_node
 
+    @contextlib.contextmanager
+    def at_root(self, at_root_hash):
+        if self.is_pruning:
```

Comment:
I'm sad because of Piper's suggestions I prefer the last two, the fluent interface is pretty!

The first one might be better though:

```
# local mutation
with trie.at_root(root_hash):
    trie.get(key)
```

If it mutates the old trie then there's no chance for users to accidentally concurrently modify a copy.

Reply:
What about inverting this and having the copy do something like acquire a lock on the parent trie, and the parent trie refuses to prune if that lock isn't free?  That should allow more freedom of movement in this API while still providing roughly equal levels of safety.

Reply:
> In particular, an optimization which modified nodes in place rather than building new ones and pruning the old ones would be allowed under the `is_pruning` interface but would break this method.

A trie that modifies in place (removing references to old nodes from the db) would certainly be "pruning" semantically, so `is_pruning` should be `True` there.

> Definitely not suggesting this PR needs to do this! But: right now there are no restrictions on self.db, a solution might involve requiring db to be something which supports snapshotting. This would remove some potential for bugs and it'd also allow us to call at_root on tries which have pruning enabled.

Yup, I am on board with doing this in future work. Having an immutable version of the trie has been on my wishlist for a long time. I listed it as a potential breaking change to make the default trie immutable: https://github.com/ethereum/py-trie/issues/85

> The first one might be better though:
> 
> ```
> # local mutation
> with trie.at_root(root_hash):
>     trie.get(key)
> ```
> If it mutates the old trie then there's no chance for users to accidentally concurrently modify a copy.

Here's my problem with that API:

```py
class TrieUser:
    def __init__(self, trie):
        self.trie = trie

    async def look_up_old_val(self, key, root):
        with trie.at_root(root):

            # this checks if nodes are missing from the trie
            while busted(trie):
                await fix_up_trie(trie)  # this is adding nodes to the trie

            # this is returning the key at whatever root hash `trie` has
            return trie[key]

    def trie_updates(self, key, val):
        # Updates might come in while awaiting above.
        # Any reference to trie now has the state root from within that context
        # even though it never invoked the context.
        # Further, the trie updates `root_hash` after setting the key & val,
        # so the returned value above is from a different root
        trie[key] = val
```